src/lib/shadow-shim-helper-rs/src/rootedcell/rc.rs

```diff
@@ -165,19 +146,81 @@ impl<T> Drop for RootedRc<T> {
     }
 }
 
-// SAFETY: RootedRc ensures that its internals can only be accessed when the
-// Root is held by the current thread, effectively synchronizing the reference
-// count.
-unsafe impl<T: Sync + Send> Send for RootedRc<T> {}
-unsafe impl<T: Sync + Send> Sync for RootedRc<T> {}
+// SAFETY: RootedRcCommon ensures that its internals can only be accessed when
+// the Root is held by the current thread, effectively synchronizing the
+// reference count.
+unsafe impl<T: Sync + Send> Send for RootedRcCommon<T> {}
+unsafe impl<T: Sync + Send> Sync for RootedRcCommon<T> {}
+
+/// Analagous to [std::rc::Rc]. In particular like [std::rc::Rc] and unlike
+/// [std::sync::Arc], it doesn't perform any atomic operations internally,
+/// making it relatively inexpensive
+///
+/// Unlike [std::rc::Rc], this type [Send] and [Sync] if `T` is. This is safe because
+/// the owner is required to prove ownership of the associated [Root]
+/// to perform any sensitive operations.
+///
+/// Instances must be destroyed using [RootedRc::safely_drop], which validates
+/// that the [Root] is held before manipulating reference counts, etc.
+/// Failing to call [RootedRc::safely_drop] results in a `panic` in debug builds,
+/// or leaking the object in release builds.
+pub struct RootedRc<T> {
+    common: RootedRcCommon<T>,
+}
+
+impl<T> RootedRc<T> {
+    /// Creates a new object associated with `root`.
+    pub fn new(root: &Root, val: T) -> Self {
+        Self {
+            common: RootedRcCommon::new(root, val),
+        }
+    }
+
+    /// Create a weak reference.
+    pub fn downgrade(&self, root: &Root) -> RootedRcWeak<T> {
+        RootedRcWeak {
+            common: self.common.clone(root, RefType::Weak),
+        }
+    }
```

Comment:
With `Rc` and `Arc`, the `downgrade` function is only accessible with the fully qualified syntax: https://doc.rust-lang.org/std/sync/struct.Arc.html#deref-behavior

Maybe do the same here for consistency?

Reply:
Makes sense; done